test/TestFullIntegration.cxx

```diff
@@ -0,0 +1,974 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The Music Player Daemon Project
+
+/**
+ * MPD Integration Test Framework
+ *
+ * A testing framework for Music Player Daemon (MPD) that provides
+ * automated setup, execution, and teardown of MPD test instances.
+ *
+ * This framework creates isolated MPD instances in temporary directories,
+ * manages their lifecycle, and provides an API for sending commands
+ * and verifying state changes.
+ */
+
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <vector>
+#include <filesystem>
+#include <cstdlib>
+#include <csignal>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <chrono>
+#include <thread>
+#include <cstring>
+#include <sstream>
+#include <stdexcept>
+#include <memory>
+#include <utility>
+#include <gtest/gtest.h>
+
+#include "fs/Path.hxx"
+#include "encoder/EncoderList.hxx"
+#include "encoder/EncoderPlugin.hxx"
+#include "encoder/EncoderInterface.hxx"
+#include "encoder/ToOutputStream.hxx"
+#include "pcm/AudioFormat.hxx"
+#include "config/Block.hxx"
+#include "io/FileOutputStream.hxx"
+
+namespace fs = std::filesystem;
+
+/**
+ * Test fixture for managing an MPD instance lifecycle.
+ *
+ * Provides management of:
+ * - Temporary test directories
+ * - MPD configuration files
+ * - MPD process lifecycle
+ * - Socket connections to MPD
+ * - State file manipulation
+ *
+ */
+class MpdTestFixture {
+private:
+	// Paths for test environment
+	fs::path test_dir_;
+	fs::path conf_path_;
+	fs::path state_path_;
+	fs::path pid_path_;
+	fs::path db_path_;
+	fs::path socket_path_;
+	fs::path music_path_;
+
+	// MPD process and connection state
+	pid_t mpd_pid_ = -1;
+	int sock_ = -1;
+	std::string mpd_executable_ = "mpd";
+	bool keep_dir_ = false;
+
+	// Constants for timing and buffer sizes
+	static constexpr int kDefaultStartupDelayMs = 1500;
+	static constexpr int kDefaultShutdownDelayMs = 10;
+	static constexpr std::size_t kSocketBufferSize = 1024;
+	static constexpr std::size_t kGreetingBufferSize = 256;
+
+	/**
+	 * Create a dummy Ogg Vorbis file.
+	 */
+	void CreateDummyOggFile(const fs::path& path) {
+		const auto *plugin = encoder_plugin_get("vorbis");
+		if (plugin == nullptr) {
+			throw std::runtime_error("Vorbis encoder plugin not found");
+		}
+
+		ConfigBlock block;
+		block.AddBlockParam("quality", "0.1", -1);
+
+		std::unique_ptr<PreparedEncoder> p_encoder(encoder_init(*plugin, block));
```

Comment:
A complicated coding style you have here. `const auto p_encoder = encoder_init(...)`.